extensions/amp-mowplayer/0.1/amp-mowplayer.js

```diff
@@ -255,20 +286,237 @@ class AmpMowplayer extends AMP.BaseElement {
+          marginLeft: 0,
+          marginRight: 0,
+          position: 'fixed',
+          top: styles.top,
+          bottom: styles.bottom,
+          left: styles.left,
+          right: styles.right,
+          zIndex: styles.zIndex,
+          width: '100%',
+          maxWidth: '500px',
+        });
+
+        if (!this.originalHeight_) {
+          this.originalHeight_ = parseInt(this.element.offsetHeight, 10);
+        }
+
+        this.attemptChangeHeight(
+          parseInt((parseInt(this.element.offsetWidth, 10) * 9) / 16, 10)
+        );
+      } else {
+        setStyles(this.element, {
+          marginLeft: previousStyles.marginLeft,
+          marginRight: previousStyles.marginRight,
+          position: 'relative',
+          top: 'auto',
+          bottom: 'auto',
+          left: 'auto',
+          right: 'auto',
+          zIndex: 'auto',
+          maxWidth: '100%',
+        });
+
+        this.attemptChangeHeight(this.originalHeight_);
+      }
+
+      this.sendMessage_('sticky_state_update', {enabled: enable});
+    };
+
+    let enabled = false;
+    const worker = () => {
+      if (this.paused_ && !enabled) {
+        return;
+      }
+
+      const rect = (enabled
+        ? fakeContainer
+        : this.element
+      ).getBoundingClientRect();
+      const height = rect.height * parseFloat(breakpoint);
+      const visible =
+        (rect.top > 0 && rect.top + height < window.innerHeight) ||
+        (rect.top < 0 && rect.top + height > 0);
+
+      if (window.mowStickyEnabled && !visible) {
+        //another sticky activated
+        return;
+      }
 
-    this.sendCommand_('listening', [
-      'amp',
-      window.location.href,
-      window.location.origin,
-      true,
-    ]);
+      if (previousVisible !== visible) {
+        enabled = !visible;
+        toggle_(enabled);
+        previousVisible = visible;
+      }
+    };
+    window.addEventListener('scroll', worker);
+    window.addEventListener('resize', worker);
+
+    this.stickyDisableWorker_ = () => {
+      toggle_(false);
+      window.removeEventListener('scroll', worker);
+      window.removeEventListener('resize', worker);
+    };
+  }
+
+  /**
+   * Check player is visible or not based on breakpoint and send message to player
+   * @private
+   * @param {Object} data
+   */
+  onVisibilityObserver_(data) {
+    let previousVisible = null;
+    const worker = () => {
+      const rect = this.element.getBoundingClientRect();
+      const height = rect.height * parseFloat(data.breakpoint);
+      const visible =
+        (rect.top > 0 && rect.top + height < window.innerHeight) ||
+        (rect.top < 0 && rect.top + height > 0);
+
+      if (previousVisible !== visible) {
+        this.sendMessage_('visibility_observer_visibility', {
+          visible,
+        });
+        previousVisible = visible;
+      }
+    };
+    window.addEventListener('scroll', worker);
+    window.addEventListener('resize', worker);
+    worker();
```

Comment:
1. `getBoundingClientRect` causes reflow. Use `element.getLayoutBox()` instead.
2. `window.innerHeight` also causes reflow. Use the `Viewport` service instead.
3. Don't attach `scroll/resize` events directly to `window. Use the `Viewport` service instead.

This also is very slow/causes jank when executed on every `scroll` event. Instead, let the frame ask for it imperatively, [like `amp-video-iframe`](https://github.com/ampproject/amphtml/blob/master/extensions/amp-video-iframe/0.1/amp-video-iframe.js#L361).

Reply:
@alanorozco Let me know if is works

```
const worker = () => {
      const {intersectionRatio} = this.element.getIntersectionChangeEntry();
      const visible = intersectionRatio > 0.5 ? true : false;
      this.sendMessage_('visibility_observer_visibility', {visible});
};
```
We also need to pause autoplay when player is less than 50%  so we need to send Message to player.